From 28102d568650ab71894fcee4694b59d945ff470b Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Sat, 10 Dec 2022 09:32:26 +0800
Subject: [PATCH 0836/1398] [Huawei] mm/mlock: Return EINVAL when overflow in
 mlock

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188105

--------------------------------

In the manual of mlock. The result of the addition addr+len was less
than addr. So return EINVAL for this due to len overflow.

Fixes: 886cc7eccc33d ("mm/mlock: return EINVAL for illegal user memory range in mlock")
Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 mm/mlock.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/mm/mlock.c b/mm/mlock.c
index 6e360939fc45..b12fdb06801e 100644
--- a/mm/mlock.c
+++ b/mm/mlock.c
@@ -649,18 +649,19 @@ static __must_check int do_mlock(unsigned long start, size_t len, vm_flags_t fla
 	unsigned long locked;
 	unsigned long lock_limit;
 	int error = -ENOMEM;
+	size_t old_len = len;
 
 	start = untagged_addr(start);
 
 	if (!can_do_mlock())
 		return -EPERM;
 
-	if (unlikely(!access_ok((void __user *)start, len)))
-		return -EINVAL;
-
 	len = PAGE_ALIGN(len + (offset_in_page(start)));
 	start &= PAGE_MASK;
 
+	if (old_len != 0 && len == 0)
+		return -EINVAL;
+
 	lock_limit = rlimit(RLIMIT_MEMLOCK);
 	lock_limit >>= PAGE_SHIFT;
 	locked = len >> PAGE_SHIFT;
@@ -715,15 +716,16 @@ SYSCALL_DEFINE3(mlock2, unsigned long, start, size_t, len, int, flags)
 SYSCALL_DEFINE2(munlock, unsigned long, start, size_t, len)
 {
 	int ret;
+	size_t old_len = len;
 
 	start = untagged_addr(start);
 
-	if (unlikely(!access_ok((void __user *)start, len)))
-		return -EINVAL;
-
 	len = PAGE_ALIGN(len + (offset_in_page(start)));
 	start &= PAGE_MASK;
 
+	if (old_len != 0 && len == 0)
+		return -EINVAL;
+
 	if (mmap_write_lock_killable(current->mm))
 		return -EINTR;
 	ret = apply_vma_lock_flags(start, len, 0);
-- 
2.17.1

